Updated Multiplayer compatibility

SokyranTheDragon opened the pull request:
I've tried to avoid making a mess in the code. Sorry if it didn't end up being the cleanest approach.

If you want me to make any changes then just let me know.

Hazzer commented on Source/ITab_Lock.cs:
```diff
@@ -222,5 +221,36 @@ private void OwnerCheckbox(Rect rect, Pawn pawn)
+            var data = LockUtility.GetData(door);
+
+            if (newState.owners == null)
+                newState.owners = data.WantedState.owners;
```
for sure you want to copy reference to list not creating a new one? Because assigning from clipboard may cause to rewrite older references

SokyranTheDragon replied:
My bad, going to fix this in a moment

Hazzer commented on Source/ITab_Lock.cs:
```diff
@@ -141,32 +143,31 @@ protected override void FillTab()
-                Data.WantedState.childLock = childLock;
-                if (Data.NeedChange)
-                    LockUtility.UpdateLockDesignation(SelDoor);
+                if (anythingChanged)
```
Any particular reason why not use an existing solution for tracking changes?

SokyranTheDragon replied:
I'm assuming you mean `LockData.NeedChange`, right? That's primarily because of how syncing works with Multiplayer.

To use `LockData.NeedChange`, the values in the object would need to be changed. If those values were set instantly on the side of the player who applied those changes, it will end in a situation where those values are different between players (until they are synced). And since those values are different for multiple people, this causes issues.

I work around setting the `anythingChanged` to true. I could alternatively compare the current `WantedState` with the values in the ITab, and apply them if there's anything that was changed. Or I could make a new, temporary `LockData` object and use that instead, but this approach seems less than ideal.

Hazzer replied:
Make sense, thanks for explanation

Hazzer commented on Source/ITab_Lock.cs:
```diff
@@ -141,32 +143,31 @@ protected override void FillTab()
             if (Clipboard.StoredState.HasValue && Widgets.ButtonText(pasteButtonsRect, "Locks_Paste".Translate()))
             {
-                Data.WantedState.CopyFrom(Clipboard.StoredState.Value);
+                SetWantedStateData(SelDoor, Clipboard.StoredState.Value);
```
It may cause reference issue.... in this case you need to create new object. Thats why there was `copyFrom` method call

SokyranTheDragon replied:
Going to fix this in a moment. Going to test the changes right now to make sure all is well.

SokyranTheDragon commented on Source/ITab_Lock.cs:
```diff
@@ -222,5 +221,36 @@ private void OwnerCheckbox(Rect rect, Pawn pawn)
+        }
+
+        [SyncMethod]
+        private static void SetWantedStateData(ThingWithComps door, LockState newState)
```
That's for simplicity when working with Multiplayer. I've made the methods static (so there's no need to make a SyncWorker for the ITab), and I'm passing as the arguments the data that's needed to apply the changes to the lock.

To stop passing the `SelDoor` as argument, I'd have to make `SelDoor` and `Data` static (and replace `SelThing` with `Find.Selector.SingleSelectedThing`, as this is what it ultimately does). The attribute would need to receive a parameter (`[SyncMethod(SyncContext.MapSelected)]`) so the data syncing would include the selected object, but that's not really an issue.

So if you'd be okay with me making the `SelDoor` and `Data` static, then I could go ahead and do it.

SokyranTheDragon commented:
Okay, I've pushed a new commit. I've fixed the list in `SetWantedStateData`, as well as I've used `CopyFrom` to copy to the current wanted state (which should fix the issue with copy/pasting).

I've also dropped the parameters and instead used the selector `LockUtility.GetData(Find.Selector.SingleSelectedThing as ThingWithComps)`. If you decide to make `SelDoor` and `Data` static, then you could replace those call with it.

Also a small fix for Multiplayer.

Let me know if there's still something you'd want me to change.

Hazzer commented on Source/ITab_Lock.cs:
```diff
@@ -222,5 +221,45 @@ private void OwnerCheckbox(Rect rect, Pawn pawn)
+        }
+
+        [SyncMethod(SyncContext.MapSelected)]
+        private static void SetWantedStateData(LockState newState)
```
I believe there is a little misunderstanding. Most of my comments were to understand what drives those changes and only issue with list required change. I'm pretty sure without that knowledge, I would refactor to use getter in the future. 😅 However, I liked more your first approach with arguments rather than selector in method. Could you bring it back for `SetWantedStateData` and `SetOwnerWantedState`? Sorry for the fuss!

Hazzer commented:
Thanks for contribution :)
